src/content.ts

```diff
@@ -1,69 +1,121 @@
+import * as batchMessages from "./core/batch/batchMesssages";
+import * as config from "./config";
+import * as messages from "./core/messages/messages";
+import * as storage from "./core/storage/storage";
+import { getRandomString, validateDSNPId, MissingUser } from "./core/utilities";
 
 /**
- * broadcast() creates a broadcast DSNP event and enqueues it for the next
- * batch. This method is not yet implemented.
+ * InvalidActivityPubOpts represents an error in the activity pub options
+ * provided to a method.
+ */
+export const InvalidActivityPubOpts = new Error("Invalid activity pub options.");
+
+/**
+ * InvalidInReplyTo represents an error in the DSNP Id provided for the
+ * inReplyTo parameter.
+ */
+export const InvalidInReplyTo = new Error("Invalid DSNP Id for inReplyTo");
+
+/**
+ * broadcast() creates an activity pub file with the given content options,
+ * uploads it with a random filename using the configured storage adapter and
+ * creates a DSNP broadcast message for the hosted file for later announcement.
+ *
+ * @throws {@link MissingSigner}
+ * Thrown if the signer is not configured.
  *
- * @param content The content for the note to broadcast
- * @param opts    Optional. Configuration overrides, such as from address, if any
- * @return The Batch Broadcast Message
+ * @param contentOptions Options for the activity pub content to generate
+ * @param opts           Optional. Configuration overrides, such as from address, if any
+ * @returns              A Signed DSNP Broadcast message ready for inclusion in a batch
  */
-export const broadcast = async (_content: ActivityPubOpts, _opts?: ConfigOpts): Promise<BatchBroadcastMessage> => {
-  throw NotImplementedError;
-
-  // const config = config.getConfig(opts);
-  //
-  // const activityPubObject = activityPub.create(activityPubOpts);
-  // const activityPubHash = activityPub.hash(activityPubObject);
-  // const uri = storage.store(activityPubObject, opts);
-  //
-  // return events.createBroadcastEvent(uri, activityPubHash);
+export const broadcast = async (
+  contentOptions: activityPub.ActivityPubOpts,
+  opts?: config.ConfigOpts
+): Promise<batchMessages.BatchBroadcastMessage> => {
+  // Create the activity pub file
+  const contentObj = activityPub.create(contentOptions);
+  if (!activityPub.validate(contentObj)) throw InvalidActivityPubOpts;
+  const content = activityPub.serialize(contentObj);
+
+  // Upload the content file
+  const filename = getRandomString();
+  const uri = await storage.put(filename, content, opts);
+
+  // Get current user id
+  const { currentUserId } = await config.getConfig(opts);
+  if (!currentUserId) throw MissingUser;
+
+  // Creates and returns the DSNP Broadcast message
+  const contentHash = activityPub.hash(contentObj);
+  const message = messages.createBroadcastMessage(currentUserId, uri.toString(), contentHash);
+
+  // Sign and return the message
+  const signedMessage = await messages.sign(message, opts);
```

Comment:
not a big deal but, what happens if this fails? should we also remove the broadcast from storage? Like a transaction, all or nothing?

Reply:
Good point, I'll bring it up in the next meeting.